__docs__/wonder-blocks-core/exports.use-latest-ref.stories.mdx

```diff
@@ -0,0 +1,95 @@
+import {useEffect, useState} from "react";
+import {Meta} from "@storybook/addon-docs";
+import {useLatestRef, View} from "@khanacademy/wonder-blocks-core";
+import Button from "@khanacademy/wonder-blocks-button";
+import {LabeledTextField} from "@khanacademy/wonder-blocks-form";
+import {Strut} from "@khanacademy/wonder-blocks-layout";
+import Spacing from "@khanacademy/wonder-blocks-spacing";
+import Color from "@khanacademy/wonder-blocks-color";
+
+<Meta
+    title="Core / Exports / useLatestRef()"
+    parameters={{
+        chromatic: {
+            disableSnapshot: true,
+        },
+    }}
+/>
+
+# useLatestRef()
+
+```ts
+function useLatestRef<T>(value: T): React.RefObject<T>;
+```
+
+The `useLatestRef` hook returns a ref that always contains the `value` passed
+to the hook during the most recent render.
+
+It can be used to wrap a possibly-changing prop in a stable value that can
+be passed to `useEffect`'s dependency array without causing unnecessary
+re-renders. For example:
+
+```ts
+export function ComponentWithDelayedAlert({message}) {
+    const messageRef = useLatestRef(message)
+    useEffect(() => {
+        const timeout = setTimeout(() => alert(messageRef.current), 5000)
+        return () => clearTimeout(timeout)
+    }, [messageRef])
+    return <p>{message}</p>
+}
+```
+
+export function ComponentWithDelayedAlert({message}) {
+    const messageRef = useLatestRef(message)
+    useEffect(() => {
+        const timeout = setTimeout(() => alert(messageRef.current), 5000)
+        return () => clearTimeout(timeout)
+    }, [messageRef])
+    return <p>{message}</p>
+}
+
+The component above will `alert()` a message five seconds after it mounts. It
+always alerts the same message it's currently rendering, even if the message
+changes between the initial mount and the alert.
+
+Here it is in action, wrapped in controls that let you change the `message` prop and remount the component.
+
+<Example1 />
+
+export function Example1() {
+    const [message, setMessage] = useState("Hello, world!")
+    const [remounts, setRemounts] = useState(0)
+    const remount = () => setRemounts(remounts + 1)
+    const unmount = () => setRemounts(0)
+    return <Context>
+        <LabeledTextField
+          label="Message"
+          description="An alert will pop up with this message 5 seconds after you click 'Remount'. You can change the message while the timer is running."
+          value={message}
+          onChange={setMessage} />
+        <Strut size={Spacing.medium_16}/>
+        <View style={{flexDirection: "row", gap: Spacing.xSmall_8}}>
+            <Button onClick={remount}>Remount</Button>
+            <Button kind="secondary" onClick={unmount}>Unmount</Button>
+        </View>
+        <Strut size={Spacing.medium_16}/>
+        <Context>
+            {remounts > 0 && <ComponentWithDelayedAlert key={remounts} message={message} />}
+        </Context>
+    </Context>
+}
```

Comment:
I wonder if the example should have something that emphasizes how the effect is not rerunning on the ref value changing. I think an example that demonstrates the value of this hook, rather than just demonstrates it working at all might be more useful.

Also, since the component is remounting over and over, I don't think the `useLatestRef` is adding any value as the whole component is tossed and remounted - i.e. there is no last ref to use. I think an example where the prop changes repeatedly and is passed through `useLatestRef` makes more sense otherwise, in the example as it stands, it isn't demonstrating the value of `useLatestRef` at all.

Reply:
I think `useLatestRef` is affecting the behavior, but I agree that the demo is pretty unmotivated, to the point where it's hard to see what the expected behavior is.

[The original motivation](https://khanacademy.slack.com/archives/C0918TZ5G/p1677878402389499) for `useLatestRef` was the need to send an analytics event when the component _unmounts_, using the props from the most recent render. I've updated the example to be closer to that. What do you think?

Reply:
Oh yeah, that's cleaner. Now it can update without being remounted but you can show how remounting would work. Long term, it'd be really cool to show the comparison of render cycles and performance of this approach versus the "traditional" deps array approach to show how just using `useIsMounted` and `useEffect` with the prop in the deps array would lead to more render cycles when doing it that way.